### PR TITLE
Add operation future

### PR DESCRIPTION
**Pull request by jonparrott**
*Note*: This adds a dependency on the `tenacity` library. It's worth it.

Towards #3617

### CODE REVIEW

File: core/google/cloud/future/operation.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+    This interacts with the long-running operations `service`_ (specific
+    to a given API) vis `HTTP/JSON`_.
+
+    .. _HTTP/JSON: https://cloud.google.com/speech/reference/rest/\
+              v1beta1/operations#Operation
+
+    Args:
+        operation (dict): Operation as a dictionary.
+        api_request (Callable): A callable used to make an API request. This
```
Comment (jonparrott) [inline core/google/cloud/future/operation.py]: @dhermes I think this is the right level to pass in here. The existing `operation` expects you to pass in a `Client` and then proceeds to access `Client._connection.api_request` which seemed beyond icky to me. Let me know if you have any problems with this being the request primitive.

File: core/setup.py [hunk 2]
```diff
@@ -57,6 +57,7 @@
     'google-auth >= 0.4.0, < 2.0.0dev',
     'google-auth-httplib2',
     'six',
+    'tenacity >= 4.0.0, <5.0.0dev'
```
Comment (dhermes) [inline core/setup.py]:
```
$ virtualenv venv
...
$ source venv/bin/activate
(venv) $ venv/bin/pip install 'tenacity >= 4.0.0, <5.0.0dev'
Collecting tenacity<5.0.0dev,>=4.0.0
...
Successfully installed ...
(venv) $ venv/bin/pip freeze
futures==3.1.1
monotonic==1.3
six==1.10.0
tenacity==4.3.0
```

File: core/google/cloud/future/base.py [hunk 3]
```diff
@@ -15,8 +15,10 @@
 """Abstract and helper bases for Future implementations."""
 
 import abc
+import concurrent.futures
```
Comment (dhermes) [inline core/google/cloud/future/base.py]: n00b question, is this 2.7 stdlib?

File: core/google/cloud/future/base.py [hunk 4]
```diff
@@ -72,8 +74,9 @@ def set_exception(self, exception):
 class PollingFuture(Future):
     """A Future that needs to poll some service to check its status.
 
-    The private :meth:`_blocking_poll` method should be implemented by
-    subclasses.
+    The private :meth:`done` method should be implemented by
```
Comment (dhermes) [inline core/google/cloud/future/base.py]: `done` doesn't seem private?

File: core/google/cloud/future/base.py [hunk 5]
```diff
@@ -89,15 +92,48 @@ def __init__(self):
...
     def _blocking_poll(self, timeout=None):
         """Poll and wait for the Future to be resolved.
 
         Args:
             timeout (int): How long to wait for the operation to complete.
                 If None, wait indefinitely.
         """
-        # pylint: disable=missing-raises
-        raise NotImplementedError()
+        if self._result_set:
```
Comment (dhermes) [inline core/google/cloud/future/base.py]: Is this "set" the verb or "set" the noun?

File: core/google/cloud/future/base.py [hunk 6]
```diff
@@ -89,15 +92,48 @@ def __init__(self):
...
+            return
+
+        retry_on = tenacity.retry_if_result(lambda result: result is not True)
```
Comment (dhermes) [inline core/google/cloud/future/base.py]: I am very :-1: on use of `lambda`, you can avoid it's use here either via:

```python
functools.partial(operator.is_not, True)
```

or

```python
def _is_not true(value):
    return value is not True
```

File: core/google/cloud/future/base.py [hunk 7]
```diff
@@ -89,15 +92,48 @@ def __init__(self):
...
+        wait_on = tenacity.wait_exponential(multiplier=1, max=10)
+
+        if timeout is not None:
```
Comment (dhermes) [inline core/google/cloud/future/base.py]: How do you feel about using `timeout is None` and swapping the blocks? (You seem to love using `not` even when you not need to.)

File: core/google/cloud/future/base.py [hunk 8]
```diff
@@ -89,15 +92,48 @@ def __init__(self):
...
             timeout (int): How long to wait for the operation to complete.
                 If None, wait indefinitely.
         """
-        # pylint: disable=missing-raises
-        raise NotImplementedError()
+        if self._result_set:
+            return
+
+        retry_on = tenacity.retry_if_result(lambda result: result is not True)
+        wait_on = tenacity.wait_exponential(multiplier=1, max=10)
```
Comment (dhermes) [inline core/google/cloud/future/base.py]: Does this support jitter? (It needs to.)

File: core/google/cloud/future/base.py [hunk 9]
```diff
@@ -113,7 +149,7 @@ def result(self, timeout=None):
             google.gax.GaxError: If the operation errors or if the timeout is
                 reached before the operation completes.
         """
-        self._blocking_poll()
+        self._blocking_poll(timeout=timeout)
```
Comment (dhermes) [inline core/google/cloud/future/base.py]: DERP

File: core/google/cloud/future/operation.py [hunk 10]
```diff
@@ -0,0 +1,250 @@
+# Copyright 2016 Google Inc.
```
Comment (dhermes) [inline core/google/cloud/future/operation.py]: 2017?

File: core/google/cloud/future/operation.py [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+
+
+class Operation(base.PollingFuture):
+    """A Future for interacting with a Google API Long-Running Operation."""
+
+    def __init__(
+            self, operation, refresh, cancel,
+            result_type, metadata_type=None):
+        """
+        Args:
```
Comment (dhermes) [inline core/google/cloud/future/operation.py]: Please move this out of `__init__` and into class docstring

File: core/google/cloud/future/operation.py [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+            metadata_type (type): The protobuf type for the operation's
+                metadata.
+        """
+        super(Operation, self).__init__()
+        self._operation = operation
+        self._refresh = refresh
+        self._cancel = cancel
+        self._result_type = result_type
+        self._metadata_type = metadata_type
+        self._completion_lock = threading.Lock()
```
Comment (dhermes) [inline core/google/cloud/future/operation.py]: Must use `threading`?

File: core/google/cloud/future/operation.py [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+        # Invoke this in case the operation came back already complete.
+        self._set_result_from_operation()
+
+    @property
+    def operation(self):
+        """google.longrunning.Operation: The current long-running operation
+        message."""
```
Comment (dhermes) [inline core/google/cloud/future/operation.py]: PEP257 (make it fit in one line, I believe in you)

File: core/google/cloud/future/operation.py [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+        return self._operation
+
+    @property
+    def metadata(self):
+        """google.protobuf.Message: the current operation metadata."""
+        if not self._operation.HasField('metadata'):
+            return None
+
+        return _helpers._from_any_pb(
```
Comment (dhermes) [inline core/google/cloud/future/operation.py]: I thought you wanted the one with `Unpack`?

File: core/google/cloud/future/operation.py [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+        """google.longrunning.Operation: The current long-running operation
+        message."""
+        return self._operation
+
+    @property
+    def metadata(self):
+        """google.protobuf.Message: the current operation metadata."""
+        if not self._operation.HasField('metadata'):
+            return None
+
```
Comment (dhermes) [inline core/google/cloud/future/operation.py]: Worth caching the output? (Probably no)

File: core/google/cloud/future/operation.py [hunk 16]
```diff
@@ -0,0 +1,250 @@
...
+        return _helpers._from_any_pb(
+            self._metadata_type, self._operation.metadata)
+
+    def _set_result_from_operation(self):
+        """Set the result or exception from the current Operation message,
+        if it is complete."""
```
Comment (dhermes) [inline core/google/cloud/future/operation.py]: PEP257, make it fit on one line

File: core/google/cloud/future/operation.py [hunk 17]
```diff
@@ -0,0 +1,250 @@
...
+                    self._result_type, self._operation.response)
+                self.set_result(response)
+            elif self._operation.HasField('error'):
+                exception = exceptions.GoogleCloudError(
+                    self._operation.error.message,
+                    errors=(self._operation.error))
+                self.set_exception(exception)
+            else:
+                exception = exceptions.GoogleCloudError(
+                    'Unknown operation error')
```
Comment (dhermes) [inline core/google/cloud/future/operation.py]: Instead:

> Unexpected state: Long-running operation had neither `response` nor `error` set.

File: core/google/cloud/future/operation.py [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+        return self._operation
+
+    def done(self):
+        """Checks to see if the operation is complete.
+
+        Returns:
+            bool: True if the operation is complete, False otherwise.
+        """
+        operation = self._refresh_and_update()
+        return operation.done
```
Comment (dhermes) [inline core/google/cloud/future/operation.py]: This is just `self._operation`, any reason to act differently?

File: core/google/cloud/future/operation.py (hunk 1 above)
Comment (dhermes) [inline core/google/cloud/future/operation.py]: No issues

File: core/google/cloud/future/operation.py [hunk 19]
```diff
@@ -0,0 +1,250 @@
...
+    This interacts with the long-running operations `service`_ (specific
+    to a given API) vis `HTTP/JSON`_.
+
+    .. _HTTP/JSON: https://cloud.google.com/speech/reference/rest/\
+              v1beta1/operations#Operation
+
+    Args:
+        operation (dict): Operation as a dictionary.
+        api_request (Callable): A callable used to make an API request. This
+            should generally be an instance of
```
Comment (dhermes) [inline core/google/cloud/future/operation.py]: "an instance of" is wrong here (i.e. it's not a class)

File: core/google/cloud/future/operation.py [hunk 20]
```diff
@@ -0,0 +1,250 @@
...
+
+
+def from_http_json(operation, api_request, result_type, **kwargs):
+    """Create an operation future from using a HTTP/JSON client.
+
+    This interacts with the long-running operations `service`_ (specific
+    to a given API) vis `HTTP/JSON`_.
+
+    .. _HTTP/JSON: https://cloud.google.com/speech/reference/rest/\
+              v1beta1/operations#Operation
```
Comment (dhermes) [inline core/google/cloud/future/operation.py]: Indent to what?

File: core/setup.py (hunk 2 above)
Comment (jonparrott) [inline core/setup.py]: Our usage should work with tenancity >= 4.0.0.

File: core/google/cloud/future/base.py (hunk 3 above)
Comment (jonparrott) [inline core/google/cloud/future/base.py]: No, but it is pulled in by tenacity.

File: core/google/cloud/future/operation.py (hunk 19 above)
Comment (jonparrott) [inline core/google/cloud/future/operation.py]: Fixed.

File: core/google/cloud/future/operation.py (hunk 20 above)
Comment (jonparrott) [inline core/google/cloud/future/operation.py]: Fixed.

File: core/google/cloud/future/base.py (hunk 4 above)
Comment (jonparrott) [inline core/google/cloud/future/base.py]: Good catch. Fixed.

File: core/google/cloud/future/operation.py [hunk 21]
```diff
@@ -94,7 +92,8 @@ def _set_result_from_operation(self):
                 self.set_exception(exception)
             else:
                 exception = exceptions.GoogleCloudError(
-                    'Unknown operation error')
+                    'Unexpected state: Long-running operation had neither '
+                    'response nor error set.')
```
Comment (dhermes) [inline core/google/cloud/future/operation.py]: I actually meant for you to include the backticks, but it's your call

**Comment by jonparrott** [comment]
@dhermes the latest release of resumable-media seems to have broken circle. Do you want me to hold off on merging this until you can resolve that? (Master will be broken by it either way).

**Comment by jonparrott** [comment]
Rebased and waiting on CI to merge.

**Pull request merged**